05-dijkstras/06-dijkstras-vs-negative-graphs.md

```diff
@@ -0,0 +1,122 @@
+
+```py
+priority queue:
+	- A
+
+visited:
+	
+
+costs/distances:
+	- A: 0 
+    - B: ∞ 
+    - C: ∞ 
+    - D: ∞
+```
+
+![Dijkstra vs.Negative Graph-3](images/negative-3.png)
+
+Step 2. Given that Node A is the only node to have a finite distance it moves to the top of our queue and we begin looking at the edges of Node A’s neighboring nodes. Here we check to see if the cost of visiting each unvisited neighbor node from the current node we are visiting is less than the estimated cost we previously noted. Here we can see that 2 and 1, for B and D respectively, are less than infinity. Therefore we will update the distances accordingly. Finally, we add our start node to the list of visited nodes so we don’t visit it again. 
+
+```py
+current: A 
+
+visited nodes: 
+    - A 
+
+priority queue: 
+ 	- A
+	- D
+	- B
+
+costs/distances: 
+    - A: 0 
+    - B: 2
+    - C: ∞ 
+    - D: 1
+```
+![Dijkstra vs.Negative Graph-4](images/negative-4.png)
+
+Step 3. The next step is to visit the node that is the closest to our start node, which is Node D due to the cost of visiting being 1, whereas Node B has a cost of 2. Here we would look at the cost of visiting all of Node D’s unvisited neighboring nodes. However, since Node D’s only neighbor is Node A, a node we’ve already visited, there are no costs to consider. 
+
+Before moving to the next step let’s recall the “greedy” behavior of Dijkstra’s algorithm. Again, Dijkstra’s Algorithm operates under the assumption that the shortest path to a destination node from a start node is an accumulation of the shortest paths connecting intermediate neighboring nodes between the starting node and destination node.Since the cost of visiting Node D is the least costly path, Dijkstra's algorithm views Node D as the most locally optimal node to visit next. This results in Node D being added to the visited list, and the algorithm will keep running with the assumption that it has found the least costly path to Node D.
+
+```py
+current: D 
+
+visited nodes: 
+    - A
+    - D 
+
+Priority queue:
+    - B
+
+costs/distances: 
+- A: 0 
+- B: 2
+- C: ∞ 
+- D: 1
+
+```
+
+![Dijkstra vs.Negative Graph-5](images/negative-5.png)
+
+Step 4. We now look at the next node at the top of our priority queue, Node B (cost of 2). We will estimate the accumulated cost of reaching Node B’s neighboring nodes, based on the cost we calculated for reaching Node B, plus the cost to reach its neighbors. Node B has a single unvisited neighbor, Node C, and the edge to reach it has a cost of -100. The -100 cost of that edge to C, added to the cost it took to reach Node B (2) results in an estimated cost to Node C of -98. We can update our cost for Node C with our new found estimate, and since we are now done visiting Node B, we can add Node B to our visited nodes. Finally, we add Node C to our priority queue.
```

Comment:
"newfound" shouldn't have a space
```suggestion
Step 4. We now look at the next node at the top of our priority queue, Node B (cost of 2). We will estimate the accumulated cost of reaching Node B’s neighboring nodes, based on the cost we calculated for reaching Node B, plus the cost to reach its neighbors. Node B has a single unvisited neighbor, Node C, and the edge to reach it has a cost of -100. The -100 cost of that edge to C, added to the cost it took to reach Node B (2) results in an estimated cost to Node C of -98. We can update our cost for Node C with our newfound estimate, and since we are now done visiting Node B, we can add Node B to our visited nodes. Finally, we add Node C to our priority queue.
```